## Breaking RSA Cryptography

$\{e , N\} =\{ 20579, 121130231\}$. $ N = p \cdot q$

We need to factor N since it is the product of two prime numbers $p$ and $q$, using Shor's algorithm. 

Using Euclid's algorithm to compute the gcd: $K = gcd(a,N)$

In [65]:
import numpy as np
import math
import random

In [66]:
def gcd(a,b):
    if a<b:
        (a,b) = (b,a)
    if (a%b) == 0:
        return b
    else:
        return (gcd(b, a%b))

In [67]:
N = 121130231

np.random.seed(1)
a = random.randint(2,N)

print(a)
print(gcd(a,N))

27754635
1


Finding the period:


In [68]:
def period(a,N):
    r = 1
    t = a
    while t != 1:
        t *= a
        t %= N
        r += 1
    return r

In [69]:
r = period(a,N)
print(r)
print(r%2)

403690
0


r is even and:
$\\ p  = gcd(a^{r/2}+1, N)
\\ q = N/p$

In [70]:
p = math.gcd(a**int(r/2)+1, N)
q = math.gcd(a**int(r/2)-1, N)
print(p,q)

15331 7901


In [71]:
p*q

121130231

which is equal to $N$
$\\ p = 15331 \\ q = 7901$

$bllhhay = (1 \times 26^5)+ (11 \times 26^4) + (7 \times 26^3) + ( 7 \times 26^2) + (0 \times 26) + 24 = 17035900$

decrypt this by finding out the private key $d$. We know the prime factors of the public key from (a). Compute multiplicative inverse to get:
$\\ d = 20579$

$ M = E^d mod N = 17035900^{20579} \ mod 121130231$

$ M = 40574 $

In [72]:
xx = 40574

x1 = int(xx//(26**5))
factor_1 = x1 * (26**5)
x2 = int((xx - factor_1)//(26**4))
factor_2 = x2 * (26**4)
x3 = int((xx - factor_1 - factor_2)//(26**3))
factor_3 = x3 * (26**3)
x4 = int((xx - factor_1 - factor_2 - factor_3)//(26**2))
factor_4 = x4 * (26**2)
x5 = int((xx - factor_1 - factor_2 - factor_3 - factor_4)//(26**1))
factor_5 = x5 * (26)
x6 = int(xx - factor_1 - factor_2 - factor_3 - factor_4 - factor_5)
display(x1,x2,x3,x4,x5,x6)

0

0

2

8

0

14

deciphering with details given in the sheet, the message is : $aaciao$